File: sip-servlets-impl/src/main/java/org/mobicents/servlet/sip/address/TelURLImpl.java
/*
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */
package org.mobicents.servlet.sip.address;

import gov.nist.javax.sip.parser.Lexer;

import java.text.ParseException;
import java.util.Iterator;

import javax.servlet.sip.TelURL;

import org.apache.log4j.Logger;
import org.mobicents.servlet.sip.SipFactories;

/**
 * Wrapper class for servlet TelURL. Would be nice if the javax.sip and servlet
 * spec adopted the same convention.
 * 
 * @author M. Ranganathan
 * 
 */
public class TelURLImpl extends URIImpl implements TelURL {

	private static Logger logger = Logger.getLogger(TelURLImpl.class
			.getCanonicalName());

	private javax.sip.address.TelURL telUrl;		

	public TelURLImpl(javax.sip.address.TelURL telUrl) {
		super(telUrl);
		this.telUrl = telUrl;
	}

	/*
	 * (non-Javadoc)
	 * @see javax.servlet.sip.TelURL#getPhoneNumber()
	 */
	public String getPhoneNumber() {
		return telUrl.getPhoneNumber();
	}

	/*
	 * (non-Javadoc)
	 * @see javax.servlet.sip.TelURL#isGlobal()
	 */
	public boolean isGlobal() {
		return telUrl.isGlobal();
	}

	/*
	 * (non-Javadoc)
	 * @see javax.servlet.sip.TelURL#setPhoneNumber(java.lang.String)
	 */
	public void setPhoneNumber(String number) {
		String phoneNumber = number;
		if(number.startsWith("+")) {			
			phoneNumber = phoneNumber.substring(1);
		}
		try {
			base_phone_number(phoneNumber);		
			telUrl.setPhoneNumber(phoneNumber);
		} catch (ParseException ex) {
			logger.error("Error setting phone number " + number);
			throw new java.lang.IllegalArgumentException("phone number " + number + " is invalid", ex);
		}
	}

	/*
	 * (non-Javadoc)
	 * @see org.mobicents.servlet.sip.address.ParameterableImpl#getParameter(java.lang.String)
	 */
	public String getParameter(String key) {
		return telUrl.getParameter(key);
	}

	/*
	 * (non-Javadoc)
	 * @see org.mobicents.servlet.sip.address.ParameterableImpl#getParameterNames()
	 */
	@SuppressWarnings("unchecked")
	public Iterator<String> getParameterNames() {
		return telUrl.getParameterNames();
	}

	/*
	 * (non-Javadoc)
	 * @see org.mobicents.servlet.sip.address.URIImpl#getScheme()
	 */
	public String getScheme() {
		return telUrl.getScheme();
	}

	/*
	 * (non-Javadoc)
	 * @see org.mobicents.servlet.sip.address.URIImpl#isSipURI()
	 */
	public boolean isSipURI() {
		return false;
	}

	/*
	 * (non-Javadoc)
	 * @see org.mobicents.servlet.sip.address.ParameterableImpl#removeParameter(java.lang.String)
	 */
	public void removeParameter(String name) {
		super.removeParameter(name);
		telUrl.removeParameter(name);
	}

	/*
	 * (non-Javadoc)
	 * @see org.mobicents.servlet.sip.address.ParameterableImpl#setParameter(java.lang.String, java.lang.String)
	 */
	public void setParameter(String name, String value) {
		super.setParameter(name, value);
		try {
			telUrl.setParameter(name, value);
		} catch (ParseException e) {
			throw new IllegalArgumentException("Problem setting parameter", e);
		}
	}

	/*
	 * (non-Javadoc)
	 * @see org.mobicents.servlet.sip.address.URIImpl#clone()
	 */
	public TelURL clone() {
		return new TelURLImpl((javax.sip.address.TelURL) this.telUrl.clone());
	}

	/*
	 * (non-Javadoc)
	 * @see org.mobicents.servlet.sip.address.URIImpl#toString()
	 */
	public String toString() {
		return this.telUrl.toString();
	}

	/*
	 * (non-Javadoc)
	 * @see javax.servlet.sip.Parameterable#getValue()
	 */
	public String getValue() {
		return this.telUrl.toString();
	}

	/*
	 * (non-Javadoc)
	 * @see javax.servlet.sip.Parameterable#setValue(java.lang.String)
	 */
	public void setValue(String value) {
		try {
			this.telUrl = SipFactories.addressFactory.createTelURL(value);
		} catch (ParseException ex) {
			throw new IllegalArgumentException("Bad url string " + value);
		}
	}

	/**
	 * {@inheritDoc}
	 */
	public String getPhoneContext() {
		return telUrl.getPhoneContext();
	}

	/**
	 * {@inheritDoc}
	 */
	public void setPhoneNumber(String number, String phoneContext) {
		setPhoneNumber(number);
		try {
			telUrl.setPhoneContext(phoneContext);
		} catch (ParseException ex) {
			logger.error("Error setting phone number " + number);
			throw new java.lang.IllegalArgumentException("phone number " + number + " is invalid", ex);
		}
	}
	
	// there might be a bug in the jsip impl in URLParser class need to check with ranga after TCK is done
	// the part in comment 
	private String base_phone_number(String number) throws ParseException {
		StringBuffer s = new StringBuffer();
		Lexer lexer = new Lexer("sip_urlLexer", number);
		
		int lc = 0;
		while (lexer.hasMoreChars()) {
			char w = lexer.lookAhead(0);
			if (Lexer.isDigit(w)
				|| w == '-'
				|| w == '.'
				|| w == '('
				|| w == ')') {
				lexer.consume(1);
				s.append(w);
				lc++;
			}
//				else if (lc > 0)
//					break;
			else
				throw new IllegalArgumentException("unexpected " + w + " in the phone number");
		}
		return s.toString();

	}
	
	
}

File: sip-servlets-examples/websocket-b2bua/src/main/sipapp/jain-sip/js/gov/nist/sip/address/TelURLImpl.js
/*
 * TeleStax, Open Source Cloud Communications  Copyright 2012. 
 * and individual contributors
 * by the @authors tag. See the copyright.txt in the distribution for a
 * full listing of individual contributors.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

/*
 *  Implementation of the JAIN-SIP  TelURLImpl.
 *  @author Yuemin Qin (yuemin.qin@orange.com)
 *  @author Laurent STRULLU (laurent.strullu@orange.com)
 *  @version 1.0 
 *  @see  gov/nist/javax/sip/address/TelURLImpl.java  
 */


function TelURLImpl() {
    if(logger!=undefined) logger.debug("TelURLImpl:TelURLImpl()");
    this.serialVersionUID = "5873527320305915954L";
    this.classname="TelURLImpl";
    this.telephoneNumber=new TelephoneNumber();
    this.scheme = "tel";
}

TelURLImpl.prototype = new GenericURI();
TelURLImpl.prototype.constructor=TelURLImpl;

TelURLImpl.prototype.setTelephoneNumber =function(telephoneNumber){
    if(logger!=undefined) logger.debug("TelURLImpl:setTelephoneNumber():telephoneNumber:"+telephoneNumber);
    this.telephoneNumber = telephoneNumber;
}

TelURLImpl.prototype.getIsdnSubAddress =function(){
    if(logger!=undefined) logger.debug("TelURLImpl:getIsdnSubAddress()");
    return this.telephoneNumber.getIsdnSubaddress();
}

TelURLImpl.prototype.getPostDial =function(){
    if(logger!=undefined) logger.debug("TelURLImpl:getPostDial()");
    return this.telephoneNumber.getPostDial();
}

TelURLImpl.prototype.getScheme =function(){
    if(logger!=undefined) logger.debug("TelURLImpl:getScheme()");
    return this.scheme;
}

TelURLImpl.prototype.isGlobal =function(){
    if(logger!=undefined) logger.debug("TelURLImpl:isGlobal()");
    return this.telephoneNumber.isGlobal();
}

TelURLImpl.prototype.isSipURI =function(){
    if(logger!=undefined) logger.debug("TelURLImpl:isSipURI()");
    return false;
}

TelURLImpl.prototype.setGlobal =function(global){
    if(logger!=undefined) logger.debug("TelURLImpl:setGlobal():global="+global);
    this.telephoneNumber.setGlobal(global);
}

TelURLImpl.prototype.setIsdnSubAddress =function(isdnSubAddress){
    if(logger!=undefined) logger.debug("TelURLImpl:setIsdnSubAddress():isdnSubAddress="+isdnSubAddress);
    this.telephoneNumber.setIsdnSubaddress(isdnSubAddress);
}

TelURLImpl.prototype.setPostDial =function(postDial){
    if(logger!=undefined) logger.debug("TelURLImpl:setPostDial():postDial="+postDial);
    this.telephoneNumber.setPostDial(postDial);
}

TelURLImpl.prototype.setPhoneNumber =function(telephoneNumber){
    if(logger!=undefined) logger.debug("TelURLImpl:setPhoneNumber():telephoneNumber="+telephoneNumber);
    this.telephoneNumber.setPhoneNumber(telephoneNumber);
} 

TelURLImpl.prototype.getPhoneNumber =function(){
    if(logger!=undefined) logger.debug("TelURLImpl:getPhoneNumber()");
    return this.telephoneNumber.getPhoneNumber();
}

TelURLImpl.prototype.toString =function(){
    if(logger!=undefined) logger.debug("TelURLImpl:toString()");
    return this.scheme + ":" + this.telephoneNumber.encode();
}

TelURLImpl.prototype.encode =function(){
    if(logger!=undefined) logger.debug("TelURLImpl:encode()");
    return this.encodeBuffer("").toString();
}

TelURLImpl.prototype.encodeBuffer =function(buffer){
    if(logger!=undefined) logger.debug("TelURLImpl:encodeBuffer():buffer="+buffer);
    buffer=buffer+this.scheme+":";
    buffer=this.telephoneNumber.encodeBuffer(buffer);
    return buffer;
}

TelURLImpl.prototype.getParameter =function(parameterName){
    if(logger!=undefined) logger.debug("TelURLImpl:getParameter():parameterName="+parameterName);
    return this.telephoneNumber.getParameter(parameterName);
}

TelURLImpl.prototype.setParameter =function(name, value){
    if(logger!=undefined) logger.debug("TelURLImpl:setParameter():name="+name+" value="+value);
    this.telephoneNumber.setParameter(name, value);
}

TelURLImpl.prototype.getParameterNames =function(){
    if(logger!=undefined) logger.debug("TelURLImpl:getParameterNames()");
    return this.telephoneNumber.getParameterNames();
}

TelURLImpl.prototype.getParameters =function(){
    if(logger!=undefined) logger.debug("TelURLImpl:getParameters()");
    return this.telephoneNumber.getParameters();
}

TelURLImpl.prototype.removeParameter =function(name){
    if(logger!=undefined) logger.debug("TelURLImpl:removeParameter():name="+name);
    this.telephoneNumber.removeParameter(name);
}

TelURLImpl.prototype.setPhoneContext =function(phoneContext){
    if(logger!=undefined) logger.debug("TelURLImpl:setPhoneContext():phoneContext="+phoneContext);
    if (phoneContext==null) {
        this.removeParameter("phone-context");
    } 
    else 
    {
        this.setParameter("phone-context",phoneContext);
    }
}

TelURLImpl.prototype.getPhoneContext =function(){
    if(logger!=undefined) logger.debug("TelURLImpl:getPhoneContext()");
    return this.getParameter("phone-context");
}

